src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -51,71 +51,129 @@ func (vmi *VMInterpreter_I) ApplyTipSetMessages(inTree st.StateTree, msgs TipSet
+
+		vmiGasUsed := message.GasLimit().Subtract(vmiGasRemaining)
+		tree = StateTree_WithGasUsedFundsTransfer(tree, vmiGasUsed, message, gasUsedFrom, gasUsedTo)
+		return tree, msg.MessageReceipt_Make(invocOutput, exitCode, vmiGasUsed)
+	}
 
-	fromActor := compTree.GetActorState(message.From())
+	_applyError := func(
+		tree st.StateTree, errExitCode exitcode.SystemErrorCode,
+		gasUsedFrom addr.Address, gasUsedTo addr.Address) (st.StateTree, msg.MessageReceipt) {
+
+		return _applyReturn(
+			tree, msg.InvocOutput_Make(nil), exitcode.SystemError(errExitCode), gasUsedFrom, gasUsedTo)
+	}
+
+	_vmiUseGas := func(amount msg.GasAmount) (vmiUseGasOK bool) {
+		vmiGasRemaining, vmiUseGasOK = vmiGasRemaining.SubtractWhileNonnegative(amount)
+		return
+	}
+
+	ok := _vmiUseGas(gascost.OnChainMessage(onChainMessageSize))
+	if !ok {
+		return _applyError(inTree, exitcode.OutOfGas, minerAddr, addr.BurntFundsActorAddr)
+	}
+
+	fromActor := inTree.GetActorState(message.From())
 	if fromActor == nil {
-		// TODO: This was originally exitcode.InvalidMethod; which is correct?
-		return inTree, _applyError(exitcode.ActorNotFound)
+		return _applyError(inTree, exitcode.ActorNotFound, minerAddr, addr.BurntFundsActorAddr)
 	}
 
-	// make sure fromActor has enough money to run the max invocation
-	maxGasCost := _gasToFIL(message.GasLimit(), message.GasPrice())
-	totalCost := message.Value() + actor.TokenAmount(maxGasCost)
+	// make sure fromActor has enough money to run with the specified gas limit
+	gasLimitCost := _gasToFIL(message.GasLimit(), message.GasPrice())
+	totalCost := message.Value() + actor.TokenAmount(gasLimitCost)
 	if fromActor.Balance() < totalCost {
-		return inTree, _applyError(exitcode.InsufficientFunds_System)
+		return _applyError(inTree, exitcode.InsufficientFunds_System, minerAddr, addr.BurntFundsActorAddr)
 	}
 
 	// make sure this is the right message order for fromActor
 	// (this is protection against replay attacks, and useful sequencing)
 	if message.CallSeqNum() != fromActor.CallSeqNum()+1 {
-		return inTree, _applyError(exitcode.InvalidCallSeqNum)
+		return _applyError(inTree, exitcode.InvalidCallSeqNum, minerAddr, addr.BurntFundsActorAddr)
 	}
 
 	// WithActorForAddress may create new account actors
-	compTree, toActor = compTree.Impl().WithActorForAddress(message.To())
+	compTreePreSend := inTree
+	compTreePreSend, toActor := compTreePreSend.Impl().WithActorForAddress(message.To())
 	if toActor == nil {
-		return inTree, _applyError(exitcode.ActorNotFound)
+		return _applyError(inTree, exitcode.ActorNotFound, message.From(), addr.BurntFundsActorAddr)
 	}
 
-	// deduct maximum expenditure gas funds first
-	compTree = _withTransferFundsAssert(compTree, message.From(), addr.BurntFundsActorAddr, maxGasCost)
+	// deduct gas limit funds from sender first
+	compTreePreSend = _withTransferFundsAssert(
+		compTreePreSend, message.From(), addr.BurntFundsActorAddr, gasLimitCost)
```

Comment:
nit: it would be more clear why this assertion is ok if you moved it up directly below where the balance is checked.

Reply:
Unfortunately I think this would cause more confusion due to the out-of-order state tree threading. But I'll add a comment explaining why the assertion is OK.